src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+﻿using System.Diagnostics.Contracts;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ProducerConsumerStream : Stream
```

Comment:
Done. The sync code paths will now wrap the async calls in Try/Catch(OperationCanceledException) and throw an ObjectDisposedException as the class is effectively disposed at that point as it's unusable any more.